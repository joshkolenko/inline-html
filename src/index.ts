import { parse } from 'node-html-parser';

import path from 'path';
import esbuild from 'esbuild';
import fs from 'fs';
import * as sass from 'sass';
import prettier from 'prettier';

import type { Config } from 'prettier';

interface Options {
  /** Attribute inline-html looks for to inline tag. Defaults to `inline` */
  attribute?: string;
  /** Directory where paths in html will resolved relative to. Defaults to `process.cwd()` */
  dir?: string;
  /** Uses _Prettier_ to format. Set to `false` for no formatting. Prettier options: https://prettier.io/docs/en/options.html */
  format?: false | Config;
  /** Paths on the filesystem that Sass will look in when resolving imports */
  loadPaths?: string[] | [];
}

/**
 * Takes a path to an HTML file and returns a string
 * with all `script` & `link` tags with attribute `inline`
 * (or user-defined attribute) replaced with compiled
 * versions of the tags' contents. Tags with `inline`
 * (or user-defined attribute) must link
 * to an external file.
 *
 * Supports `sass` & `ts`
 *
 * @example
 *
 * ```js
 * import inlineHTML from 'inline-html'
 *
 * const html = await inlineHTML('path/to/file')
 * console.log(html)
 * ```
 *
 * @param p Path to HTML file
 * @param options Options object, set custom attribute, sass `loadPaths` & format settings
 * @returns
 */
export const inlineHTML = async (p: string, options?: Options) => {
  const config: Options = {
    attribute: 'inline',
    loadPaths: [],
    format: {
      printWidth: 200,
      tabWidth: 2,
      semi: true,
    },
    ...options,
  };

  let html: string, dir: string;

  if (fs.existsSync(p) && fs.lstatSync(p).isFile()) {
    html = fs.readFileSync(p, 'utf-8');
    dir = path.parse(p).dir;
  } else {
    html = p;
    dir = options?.dir || process.cwd();
  }

  const document = parse(html, {
    comment: true,
  });

  const nodes = Array.from(document.querySelectorAll(`[${config.attribute}]`));

  for (const node of nodes) {
    if (node.tagName === 'LINK') {
      const filePath = path.resolve(dir, node.getAttribute('href') || '');

      if (!fs.existsSync(filePath)) {
        throw new Error(`File not found at: ${filePath}`);
      }

      const output = sass.compile(filePath, {
        charset: false,
        loadPaths: config.loadPaths,
      }).css;

      node.replaceWith(`<style>\n${output}\n</style>`);
    }

    if (node.tagName === 'SCRIPT') {
      const filePath = path.resolve(dir, node.getAttribute('src') || '');

      if (!fs.existsSync(filePath)) {
        throw new Error(`File not found at: ${filePath}`);
      }

      const result = await esbuild.build({
        entryPoints: [filePath],
        format: 'iife',
        bundle: true,
        write: false,
      });

      const output = result.outputFiles[0].text;

      node.replaceWith(`<script>\n${output}</script>`);
    }
  }

  if (!config.format) {
    return document.toString();
  }

  const formatted = prettier.format(document.toString(), {
    parser: 'html',
    ...config.format,
  });

  return formatted;
};
