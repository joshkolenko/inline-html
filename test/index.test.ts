import { expect, describe, test } from 'vitest';

import fs from 'fs';
import path from 'path';

import { inlineHTML, type Options } from '../src';

const baseOptions: Options = { loadPaths: ['test/fixtures/scss/load/path/'] };
const fileOptions: Options = { ...baseOptions };
const stringOptions: Options = { ...baseOptions, dir: 'test/fixtures/' };

const filePath = path.resolve('test/fixtures/index.html');
const fileContents = fs.readFileSync(filePath, 'utf-8');

describe('inline', () => {
  test('file', async () => {
    const filePath = path.resolve('test/fixtures/html/index.html');
    const html = await inlineHTML(filePath, fileOptions);
    expect(html).toMatchSnapshot();
  });

  test('string', async () => {
    const html = await inlineHTML(fileContents, stringOptions);
    expect(html).toMatchSnapshot();
  });
});

describe('compile & inline', () => {
  test('css', async () => {
    const html = `<link rel="stylesheet" href="css/main.css" inline />`;
    const result = await inlineHTML(html, stringOptions);
    expect(result).toMatchSnapshot();
  });

  test('scss', async () => {
    const html = `<link rel="stylesheet" href="scss/main.scss" inline />`;
    const result = await inlineHTML(html, stringOptions);
    expect(result).toMatchSnapshot();
  });

  test('js', async () => {
    const html = `<script src="js/index.js" inline></script>`;
    const result = await inlineHTML(html, stringOptions);
    expect(result).toMatchSnapshot();
  });

  test('ts', async () => {
    const html = `<script src="ts/index.ts" inline></script>`;
    const result = await inlineHTML(html, stringOptions);
    expect(result).toMatchSnapshot();
  });
});

describe('options', () => {
  test('attribute', async () => {
    const html = `<link rel="stylesheet" href="css/main.css" custom-attribute />`;
    const result = await inlineHTML(html, { ...stringOptions, attribute: 'custom-attribute' });
    expect(result).toMatchSnapshot();
  });

  test('disable css transform', async () => {
    const result = await inlineHTML(filePath, { ...fileOptions, transformCss: false });
    expect(result).toMatchSnapshot();
  });

  test('minify css', async () => {
    const result = await inlineHTML(filePath, { ...fileOptions, minifyCss: true });
    expect(result).toMatchSnapshot();
  });

  test('minify js', async () => {
    const result = await inlineHTML(filePath, { ...fileOptions, minifyJs: true });
    expect(result).toMatchSnapshot();
  });
});
