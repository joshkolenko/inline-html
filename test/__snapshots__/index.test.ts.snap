// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compile & inline > css 1`] = `
"<style>
body {
  margin: 0;
}

body div {
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
  display: -ms-flexbox;
  display: flex;
}

h1 {
  color: red;
}

p {
  color: #00f;
}
</style>"
`;

exports[`compile & inline > js 1`] = `
"<script>
(() => {
  // test/fixtures/js/module.js
  var message = "Hello, world!";
  var module = () => {
    return message;
  };

  // test/fixtures/js/index.js
  var result = module(message);
  console.log(result);
})();
</script>"
`;

exports[`compile & inline > scss 1`] = `
"<style>
.partial {
  color: green;
}

.partial div {
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
  display: -ms-flexbox;
  display: flex;
}

body {
  margin: 0;
}

body h1 {
  color: red;
}

body p {
  color: #00f;
}

.load-path-test {
  color: red;
}
</style>"
`;

exports[`compile & inline > ts 1`] = `
"<script>
(() => {
  // test/fixtures/ts/module.ts
  var message = "Hello, world!";
  var module = () => {
    return message;
  };

  // test/fixtures/ts/index.ts
  var result = module(message);
  console.log(result);
})();
</script>"
`;

exports[`inline > file 1`] = `
"<!-- css -->
<style>
body {
  margin: 0;
}

body div {
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
  display: -ms-flexbox;
  display: flex;
}

h1 {
  color: red;
}

p {
  color: #00f;
}
</style>

<!-- scss -->
<style>
.partial {
  color: green;
}

.partial div {
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
  display: -ms-flexbox;
  display: flex;
}

body {
  margin: 0;
}

body h1 {
  color: red;
}

body p {
  color: #00f;
}

.load-path-test {
  color: red;
}
</style>

<!-- js -->
<script>
(() => {
  // test/fixtures/js/module.js
  var message = "Hello, world!";
  var module = () => {
    return message;
  };

  // test/fixtures/js/index.js
  var result = module(message);
  console.log(result);
})();
</script>

<!-- ts -->
<script>
(() => {
  // test/fixtures/ts/module.ts
  var message = "Hello, world!";
  var module = () => {
    return message;
  };

  // test/fixtures/ts/index.ts
  var result = module(message);
  console.log(result);
})();
</script>
"
`;

exports[`inline > string 1`] = `
"<!-- css -->
<style>
body {
  margin: 0;
}

body div {
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
  display: -ms-flexbox;
  display: flex;
}

h1 {
  color: red;
}

p {
  color: #00f;
}
</style>

<!-- scss -->
<style>
.partial {
  color: green;
}

.partial div {
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
  display: -ms-flexbox;
  display: flex;
}

body {
  margin: 0;
}

body h1 {
  color: red;
}

body p {
  color: #00f;
}

.load-path-test {
  color: red;
}
</style>

<!-- js -->
<script>
(() => {
  // test/fixtures/js/module.js
  var message = "Hello, world!";
  var module = () => {
    return message;
  };

  // test/fixtures/js/index.js
  var result = module(message);
  console.log(result);
})();
</script>

<!-- ts -->
<script>
(() => {
  // test/fixtures/ts/module.ts
  var message = "Hello, world!";
  var module = () => {
    return message;
  };

  // test/fixtures/ts/index.ts
  var result = module(message);
  console.log(result);
})();
</script>
"
`;

exports[`options > attribute 1`] = `
"<style>
body {
  margin: 0;
}

body div {
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
  display: -ms-flexbox;
  display: flex;
}

h1 {
  color: red;
}

p {
  color: #00f;
}
</style>"
`;

exports[`options > disable css transform 1`] = `
"<!-- css -->
<style>
body {
  margin: 0;

  div {
    display: flex;
    place-items: center;
  }
}

h1 {
  color: red;
}

p {
  color: blue;
}
</style>

<!-- scss -->
<style>
.partial {
  color: green;
}
.partial div {
  display: flex;
  place-items: center;
}

body {
  margin: 0;
}
body h1 {
  color: red;
}
body p {
  color: blue;
}

.load-path-test {
  color: red;
}
</style>

<!-- js -->
<script>
(() => {
  // test/fixtures/js/module.js
  var message = "Hello, world!";
  var module = () => {
    return message;
  };

  // test/fixtures/js/index.js
  var result = module(message);
  console.log(result);
})();
</script>

<!-- ts -->
<script>
(() => {
  // test/fixtures/ts/module.ts
  var message = "Hello, world!";
  var module = () => {
    return message;
  };

  // test/fixtures/ts/index.ts
  var result = module(message);
  console.log(result);
})();
</script>
"
`;

exports[`options > minify css 1`] = `
"<!-- css -->
<style>
body{margin:0}body div{-ms-flex-align:center;align-items:center;justify-items:center;display:-ms-flexbox;display:flex}h1{color:red}p{color:#00f}
</style>

<!-- scss -->
<style>
.partial{color:green}.partial div{-ms-flex-align:center;align-items:center;justify-items:center;display:-ms-flexbox;display:flex}body{margin:0}body h1{color:red}body p{color:#00f}.load-path-test{color:red}
</style>

<!-- js -->
<script>
(() => {
  // test/fixtures/js/module.js
  var message = "Hello, world!";
  var module = () => {
    return message;
  };

  // test/fixtures/js/index.js
  var result = module(message);
  console.log(result);
})();
</script>

<!-- ts -->
<script>
(() => {
  // test/fixtures/ts/module.ts
  var message = "Hello, world!";
  var module = () => {
    return message;
  };

  // test/fixtures/ts/index.ts
  var result = module(message);
  console.log(result);
})();
</script>
"
`;

exports[`options > minify js 1`] = `
"<!-- css -->
<style>
body {
  margin: 0;
}

body div {
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
  display: -ms-flexbox;
  display: flex;
}

h1 {
  color: red;
}

p {
  color: #00f;
}
</style>

<!-- scss -->
<style>
.partial {
  color: green;
}

.partial div {
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
  display: -ms-flexbox;
  display: flex;
}

body {
  margin: 0;
}

body h1 {
  color: red;
}

body p {
  color: #00f;
}

.load-path-test {
  color: red;
}
</style>

<!-- js -->
<script>
(()=>{var o="Hello, world!",e=()=>o;var s=e(o);console.log(s);})();
</script>

<!-- ts -->
<script>
(()=>{var e="Hello, world!",o=()=>e;var s=o(e);console.log(s);})();
</script>
"
`;
